libraries/standard/mqtt/src/mqtt_state.c

```diff
@@ -0,0 +1,588 @@
+    /* There are more QoS2 cases than QoS1, so initialize to that. */
+    bool qosValid = ( qos == MQTTQoS2 );
+
+    switch( packetType )
+    {
+        case MQTTPuback:
+            qosValid = ( qos == MQTTQoS1 );
+            calculatedState = MQTTPublishDone;
+            break;
+        case MQTTPubrec:
+            /* Incoming publish: send PUBREC, PUBREL pending.
+             * Outgoing publish: receive PUBREC, send PUBREL. */
+            calculatedState = ( opType == MQTT_SEND )? MQTTPubRelPending : MQTTPubRelSend;
+            break;
+        case MQTTPubrel:
+            /* Incoming publish: receive PUBREL, send PUBCOMP.
+             * Outgoing publish: send PUBREL, PUBCOMP pending. */
+            calculatedState = ( opType == MQTT_SEND )? MQTTPubCompPending : MQTTPubCompSend;
+            break;
+        case MQTTPubcomp:
+            calculatedState = MQTTPublishDone;
+            break;
+        default:
+            /* No other ack type. */
+            break;
+    }
+    /* Sanity check, make sure ack and QoS agree. */
+    if( !qosValid )
+    {
+        calculatedState = MQTTStateNull;
+    }
+    return calculatedState;
+}
+
+MQTTPublishState_t MQTT_UpdateStateAck( MQTTContext_t * pMqttContext,
+                                        uint16_t packetId,
+                                        MQTTPubAckType_t packetType,
+                                        MQTTStateOperation_t opType )
+{
+    MQTTPublishState_t newState = MQTTStateNull;
+    MQTTPublishState_t currentState = MQTTStateNull;
+    bool isPublishOutgoing = _isPublishOutgoing( packetType, opType );
+    bool shouldDeleteRecord = false;
+    bool isTransitionValid = false;
+    MQTTQoS_t qos = MQTTQoS0;
+    size_t recordIndex = MQTT_STATE_ARRAY_MAX_COUNT;
+    MQTTPubAckInfo_t * records = NULL;
+
+    if( isPublishOutgoing )
+    {
+        records = pMqttContext->outgoingPublishRecords;
+    }
+    else
+    {
+        records = pMqttContext->incomingPublishRecords;
+    }
+    recordIndex = _findInRecord( records,
+                                 MQTT_STATE_ARRAY_MAX_COUNT,
+                                 packetId,
+                                 &qos,
+                                 &currentState );
+
+    if( recordIndex < MQTT_STATE_ARRAY_MAX_COUNT )
+    {
+        newState = MQTT_CalculateStateAck( packetType, opType, qos );
+        shouldDeleteRecord = ( newState == MQTTPublishDone );
+        isTransitionValid = _validateTransitionAck( currentState, newState );
+        if( isTransitionValid )
+        {
+            _updateRecord( records,
+                           MQTT_STATE_ARRAY_MAX_COUNT,
+                           recordIndex,
+                           newState,
+                           shouldDeleteRecord );
+        }
+        else
+        {
+            newState = MQTTStateNull;
+        }
+        
+    }
+    return newState;
+}
+
+uint16_t MQTT_StateSelect( MQTTContext_t * pMqttContext,
```

Comment:
Seem to be missing the cursor.  You can use an in-out style argument for your recordIndex - just be sure to increment in the success case.

Reply:
Added cursor argument and type